### ISSUE TITLE
New line issues with Karaf Console command output when using SSH session

### ISSUE THREAD

**Issue by wborn** [openhab/openhab-distro#764]
When interacting with the Karaf Console through a SSH session there are new line issues with the output of console commands (most of the time). 

This issue occurs with openHAB 2.4.0-SNAPSHOT (build # 1363).

# Command examples

![selection_013](https://user-images.githubusercontent.com/12213581/45587851-d985bf00-b90b-11e8-86f1-7016a262281d.png)

# Logging examples

![selection_014](https://user-images.githubusercontent.com/12213581/45587853-dc80af80-b90b-11e8-815c-f4909d7f736e.png)

It is probably a regression caused by the upgrade to Karaf 4.2.1 (#761). I already tried downgrading Karaf dependencies like jansi and sshd-core but that didn't help. I haven't found a known issue in the Karaf issue tracker for this.

Unaffected are the output of the "openHAB" welcome message, `help`  and command completion suggestions.

I also made sure it's not caused by #763 by removing `shell.colors.script`.

**Comment by wborn** [comment]
Seems we are not the only ones with this issue:

http://karaf.922171.n3.nabble.com/Strange-behaviour-when-accessing-console-via-ssh-client-into-a-dockernized-Karaf-td4053844.html

**Comment by wborn** [comment]
There's now a Karaf issue for this:

[KARAF-5916](https://issues.apache.org/jira/browse/KARAF-5916): ssh terminal rendering is not correct using ssh client

**Comment by tarcus69** [comment]
I have also hit this issue, am using your suggested workaround by logging in directly to the openhab box (which is what I always did anyway) and run:

$OPENHAB_RUNTIME/bin/client

**Comment by lsafelix75** [comment]
just run openhab-cli

**Comment by yveshanoulle** [comment]
Is there a workaround from another machine? 
the openhab machine is in my attic, and I have no screen attached to this machine.

**Comment by wborn** [comment]
You can use the Apache Karaf `client` that comes with openHAB to also connect to other Karaf instances (e.g. openHAB running on another machine):

```
$ runtime/bin/client --help
Apache Karaf client
  -a [port]     specify the port to connect to
  -h [host]     specify the host to connect to
  -u [user]     specify the user name
  -p [password] specify the password (optional, if not provided, the password is prompted)
  --help        shows this help message
  -v            raise verbosity
  -l            set client logging level. Set to 0 for ERROR logging and up to 4 for TRACE
  -r [attempts] retry connection establishment (up to attempts times)
  -d [delay]    intra-retry delay (defaults to 2 seconds)
  -b            batch mode, specify multiple commands via standard input
  -f [file]     read commands from the specified file
  -k [keyFile]  specify the private keyFile location when using key login, need have BouncyCastle registered as security provider using this flag
  -t [timeout]  define the client idle timeout
  [commands] [--]   commands to run
If no commands are specified, the client will be put in an interactive mode
```

It's also part of the .zip and .tar.gz distributions so you don't have to do an apt install to use it.

**Comment by yveshanoulle** [comment]
Do I understand correctly, that for that I would need to install openhab on another machine?
If so I assume I have to make sure only one instance is actually running?

**Comment by wborn** [comment]
You can untar/unzip the archive and then just run the `client` from the extracted files with the right command line options. So there is no need for running openHAB on the machine from which you are connecting.

**Comment by yveshanoulle** [comment]
Where can I find theses archives? 
I want to look at the logging from my mac. 
Openhab it's self is running on an Pine64, installed using apt-get .

**Comment by BClark09** [comment]
Alternatively, you could just SSH into the Pine64 first, then run `openhab-cli console` or `${OPENHAB_RUNTIME}/bin/client`. I believe Pine64 images have ssh enabled by default and a Mac's terminal has an SSH client:

```
ssh <ip>
<username when prompted>
<password when prompted>
openhab-cli console
```

**Comment by yveshanoulle** [comment]
> openhab-cli console
ah so that also  works SSH.
I thought that was only on the machine. 

Then what's the difference between this and ssh -p 8101 openhab@localhost ?

**Comment by BClark09** [comment]
SSH is a protocol that allows you to remotely use a machine, just as if you were connected to it using a screen and keyboard.

By using SSH for the machine, you connect to that device and can use any commands you would normally use via a keyboard.

openHAB (which uses Karaf) allows SSH connections too, but some operating systems don't have an included SSH client (namely windows, but I think there are others). Karaf and openHAB come packaged with its own and because they are known, can also bypass the need to use a password.

`openhab-cli console` is simply a shortcut that runs the client included with Karaf.

**Comment by yveshanoulle** [comment]
I know SSH, I was using something else before karaf (I don't remember what anymore) yet that stopped working when karaf was added (I think in 2.0) 
From the first mention of "openhab-cli console" it mention to do that on the machine itself, so I assumed (wrongly) that it would not be able to do that over SSH .

**Comment by wborn** [comment]
This pesky bug finally got fixed in Karaf with https://github.com/apache/karaf/pull/635.
So I've build a [Karaf 4.2.1-sp1 release](https://github.com/wborn/karaf/releases/tag/karaf-4.2.1-sp1) with just this fix.

You can apply/test this fix on a openHAB 2.4.0 instance by:
1. Download [org.apache.karaf.shell.ssh-4.2.1-sp1.jar](https://github.com/wborn/karaf/releases/download/karaf-4.2.1-sp1/org.apache.karaf.shell.ssh-4.2.1-sp1.jar)
2. Stop openHAB
3. Copy `org.apache.karaf.shell.ssh-4.2.1-sp1.jar` over 
`$OPENHAB_HOME/runtime/system/org/apache/karaf/shell/org.apache.karaf.shell.ssh/4.2.1/org.apache.karaf.shell.ssh-4.2.1.jar`
4. Empty cache (remove files) in `$OPENHAB_USERDATA/cache`
5. Start openHAB

We can either integrate this fix with some overrides or with the scheduled Karaf 4.2.2 release ([ETA 23 Nov 2018](https://karaf.apache.org/download.html)). What would you prefer @kaikreuzer?

**Comment by kaikreuzer** [comment]
From my experience, Karaf releases tend to be delayed (not blaming them, at least they have an ETA ;-)), so it probably might make sense to patch it for the time being.
@wborn Are you familiar enough with the override mechanism and would dare to come up with an PR?
